tests/integration-tests/tests/integration/chained_authority_tests.rs

```diff
@@ -0,0 +1,363 @@
+        Name::from_ascii("example.com.").unwrap(),
+        sec_lookup_records,
+        sec_consult_records,
+    );
+
+    catalog.upsert(
+        primary_authority.origin().clone(),
+        vec![Arc::new(primary_authority), Arc::new(secondary_authority)],
+    );
+
+    // First test - the record only exists in the primary authority
+    basic_test(&catalog, "primaryonly.example.com.", pri_lookup_ip).await;
+
+    // Second test -- the record exists in both authorities; confirm the primary authority data
+    // is returned
+    basic_test(&catalog, "continueboth.example.com.", pri_lookup_ip).await;
+
+    // Third test -- the record exists in the primary authority, but is overwritten by a record in
+    // the secondary authority
+    basic_test(&catalog, "overwrite.example.com.", sec_consult_ip).await;
+
+    // Fourth test -- the record exists in the primary authority and is returned with Break -
+    // verify consult methods are not consulted for any authority.
+    basic_test(&catalog, "breakok.example.com.", pri_lookup_ip).await;
+
+    // Fifth test -- primary returns skip, and the second authority has the record - verify the
+    // rdata from the secondary authority is returned.
+    basic_test(&catalog, "skipprimary.example.com.", sec_lookup_ip).await;
+
+    // Sixth test - both authorities skip.  Verify the catalog returns Servfail
+    error_test(&catalog, "skipboth.example.com.", ResponseCode::ServFail).await;
+
+    // Seventh test -- Primary returns Continue(Err), secondary returns Ok with a record
+    basic_test(&catalog, "primaryerr.example.com.", sec_consult_ip).await;
+
+    // Eighth test -- Primary returns Break(Err); secondary consult WOULD result in a record
+    // returned; verify no records
+    error_test(&catalog, "breakerr.example.com.", ResponseCode::NoError).await;
+}
+
+#[derive(Debug)]
+enum ResponseType {
+    ContinueOk,
+    BreakOk,
+    ContinueErr,
+    BreakErr,
+    Skip,
+}
+
+/// This is a lookup table for inner_lookup, which is called by the test authority lookup and
+/// consult methods.  Each entry in the Vec is a tuple, which represent a query string and action
+/// pair.  The action is wrapped in an Option - for None variants, if the lookup or consult method
+/// is queried for that name, the test will panic.  This covers cases where lookup and/or consult
+/// should not be called, such as verifying that LookupControlFlow::Break is working properly.
+/// The Some variant will include a ResponseType and a record. ResponseType is an enum that maps 1:1
+/// to LookupControlFlow, and controls the control flow type returned to the catalog.  The record is
+/// always an A record, and will be returned with the lookup records in Continue(Ok) and Break(Ok)
+/// responses. It is used to distinguish between the primary and secondary authority having been the
+/// source of the answer returned by the catalog.
+type TestRecords = Vec<(&'static str, Option<(ResponseType, A)>)>;
```

Comment:
Suggest moving `ResponseType` and `TestRecords` down below `TestAuthority` to get a cleaner top-down order (instead of the topologically inconsistent `chained_authority_test`, then `RecordType`, then `TestRecords`, then `TestAuthority` order).